src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+    assume u f hf,
+    change f ∈ {local_homeomorph.refl H} ∪ {e : local_homeomorph H H | e.source = ∅} at hf,
+    simp only [singleton_union, mem_set_of_eq, mem_insert_iff] at hf,
+    cases hf,
+    { rw hf,
+      apply u.id_mem },
+    { apply u.locality,
+      assume x hx,
+      rw [hf, mem_empty_eq] at hx,
+      exact hx.elim }
+  end,
+  ..structure_groupoid.partial_order }
+
+/-- To construct a groupoid, one may consider classes of local homeos such that both the function
+and its inverse have some property. If this property is stable under composition,
+one gets a groupoid. `pregroupoid` bundles the properties needed for this construction, with the
+groupoid of smooth functions with smooth inverses as an application. -/
+structure pregroupoid (H : Type*) [topological_space H] :=
+(P        : (H → H) → (set H) → Prop)
+(comp     : ∀{f g u v}, P f u → P g v → is_open (u ∩ f ⁻¹' v) → P (g ∘ f) (u ∩ f ⁻¹' v))
+(id_mem   : P id univ)
+(locality : ∀{f u}, is_open u → (∀x∈u, ∃v, is_open v ∧ x ∈ v ∧ P f (u ∩ v)) → P f u)
+(congr    : ∀{f g : H → H} {u}, is_open u → (∀x∈u, g x = f x) → P f u → P g u)
+
+/-- Construct a groupoid of local homeos for which the map and its inverse have some property,
+from a pregroupoid asserting that this property is stable under composition. -/
+def groupoid_of_pregroupoid (PG : pregroupoid H) : structure_groupoid H :=
+{ members  := {e : local_homeomorph H H | PG.P e.to_fun e.source ∧ PG.P e.inv_fun e.target},
+  comp     := λe e' he he', begin
+    split,
+    { apply PG.comp he.1 he'.1,
+      apply e.continuous_to_fun.preimage_open_of_open e.open_source e'.open_source },
+    { apply PG.comp he'.2 he.2,
+      apply e'.continuous_inv_fun.preimage_open_of_open e'.open_target e.open_target }
+  end,
+  inv      := λe he, ⟨he.2, he.1⟩,
+  id_mem   := ⟨PG.id_mem, PG.id_mem⟩,
+  locality := λe he, begin
+    split,
+    { apply PG.locality e.open_source (λx xu, _),
+      rcases he x xu with ⟨s, s_open, xs, hs⟩,
+      refine ⟨s, s_open, xs, _⟩,
+      convert hs.1,
+      exact (interior_eq_of_open s_open).symm },
+    { apply PG.locality e.open_target (λx xu, _),
+      rcases he (e.inv_fun x) (e.map_target xu) with ⟨s, s_open, xs, hs⟩,
+      refine ⟨e.target ∩ e.inv_fun ⁻¹' s, _, ⟨xu, xs⟩, _⟩,
+      { exact continuous_on.preimage_open_of_open e.continuous_inv_fun e.open_target s_open },
+      { rw [← inter_assoc, inter_self],
+        convert hs.2,
+        exact (interior_eq_of_open s_open).symm } },
+  end,
+  eq_on_source := λe e' he ee', begin
+    split,
+    { apply PG.congr e'.open_source ee'.2,
+      have Z := ee'.1,
+      erw ee'.1,
+      exact he.1 },
+    { have A := eq_on_source_symm ee',
+      apply PG.congr e'.symm.open_source A.2,
+      convert he.2,
+      rw A.1,
+      refl }
+  end }
+
+lemma mem_groupoid_of_pregroupoid (PG : pregroupoid H) (e : local_homeomorph H H) :
+  e ∈ groupoid_of_pregroupoid PG ↔ PG.P e.to_fun e.source ∧ PG.P e.inv_fun e.target :=
+by refl
```

Comment:
Does it make a difference to use `iff.rfl` here? I now using `rfl` puts an extra refl attribute on a lemma.